test/unit/org/apache/cassandra/repair/FuzzTestBase.java

```diff
@@ -0,0 +1,1306 @@
+                            failures.add(new AssertionError(event.getMessage()));
+                    });
+                }
+                return repair;
+            }
+
+            public RangesAtEndpoint getLocalReplicas(String ks)
+            {
+                return StorageService.instance.getReplicas(ks, broadcastAddressAndPort());
+            }
+
+            public Collection<? extends Range<Token>> getPrimaryRanges(String ks)
+            {
+                return StorageService.instance.getPrimaryRangesForEndpoint(ks, broadcastAddressAndPort());
+            }
+
+            public Collection<? extends Range<Token>> getPrimaryRangesWithinDC(String ks)
+            {
+                return StorageService.instance.getPrimaryRangeForEndpointWithinDC(ks, broadcastAddressAndPort());
+            }
+
+            @Override
+            public ActiveRepairService repair()
+            {
+                return activeRepairService;
+            }
+
+            @Override
+            public IValidationManager validationManager()
+            {
+                return validationManager;
+            }
+
+            @Override
+            public TableRepairManager repairManager(ColumnFamilyStore store)
+            {
+                return new CassandraTableRepairManager(store, this)
+                {
+                    @Override
+                    public void snapshot(String name, Collection<Range<Token>> ranges, boolean force)
+                    {
+                        // no-op
+                    }
+                };
+            }
+
+            @Override
+            public StreamExecutor streamExecutor()
+            {
+                return streamExecutor;
+            }
+        }
+
+        private Message serde(Message msg)
+        {
+            try (DataOutputBuffer b = DataOutputBuffer.scratchBuffer.get())
+            {
+                int messagingVersion = MessagingService.Version.CURRENT.value;
+                Message.serializer.serialize(msg, b, messagingVersion);
+                DataInputBuffer in = new DataInputBuffer(b.unsafeGetBufferAndFlip(), false);
+                return Message.serializer.deserialize(in, msg.from(), messagingVersion);
+            }
+            catch (Throwable e)
+            {
+                failures.add(e);
+                return null;
+            }
+        }
+    }
+
+    private static <T> Gen<T> fromQT(org.quicktheories.core.Gen<T> qt)
+    {
+        return rs -> {
+            JavaRandom r = new JavaRandom(rs.asJdkRandom());
+            return qt.generate(r);
+        };
+    }
+
+    public static class HackStrat extends LocalStrategy
+    {
+        public HackStrat(String keyspaceName, TokenMetadata tokenMetadata, IEndpointSnitch snitch, Map<String, String> configOptions)
+        {
+            super(keyspaceName, tokenMetadata, snitch, configOptions);
+        }
+    }
+
+    public static class ClockAccess implements Clock
```

Comment:
added

```
+    /**
+     * Since the clock is accessable via {@link Global#currentTimeMillis()} and {@link Global#nanoTime()}, and only repair subsystem has the requirement not to touch that, this class
+     * acts as a safty check that validates that repair does not touch these methods but allows the other subsystems to do so.
+     */
```